### PR TITLE
Add fix for 1.15 BER crash

### PR DESCRIPTION
**Pull request by Draylar**
In 1.15 snapshots, `BlockEntityRender` registration was changed a small bit. You now have to access `BlockEntityRenderer.INSTANCE` and pass it up in your BER class.

This has a small side effect: when mods reference the `INSTANCE` field, the Piston BER is initialized. This happens _before_ `MinecraftClient`'s instance of `BlockRenderManager` is initialized, and the piston renderer class saves said instance on creation. 

This means: when pistons attempt to render after being powered, they will crash the game, assuming another mod is present that registers a BER.

This patch simply double-checks the manager in the piston class isn't null before using it (and initializes it if it's not). As far as I can see, there aren't any other usages that would break under the same circumstances.

### CODE REVIEW

**Comment by liach** [comment]
Then I'd suggest restoring the lazy access to block entity render dispatcher like in the old 1.14 api.

**Comment by Draylar** [comment]
After discussing the issue with a few people in #api, I came up with a simple system similar to what liach suggested. The key change is that `BlockEntityRendererRegistry#register` now takes a `Function<BlockEntityRenderDispatcher, BlockEntityRenderer<E>>` instead of a direct instance of a BER. The function is then applied inside `BlockEntityRenderDispatcher` at the end of init to ensure `INSTANCE` isn't initialized until needed.

Our current system supports registering renderers post-init, and the thought came up on how to continue to support that. The new register method checks a boolean which is set at the end of `BlockEntityRenderDispatcher#init`, and if it's true (BERs have been registered), we directly add the renderer to the renderers map instead of storing it for later. 

I've opted to keep the piston mixin; someone will eventually access the `INSTANCE` field and accidentally initialize it early.

Naming for fields and methods isn't final; I just wanted to get feedback before adding docs and such.

**Questions**:
- I currently reference `BlockEntityRenderDispatcher.INSTANCE` on the `else` side of an if-statement. Assuming the else side doesn't trigger, would `INSTANCE` ever be initialized (just due to referencing it)? I would guess it doesn't, but I figure I should ask.
- There are currently 2 mixins used for `BlockEntityRenderDispatcher`; one is for injects, and one is for an invoker. Is this fine, or should I swap to a mixin interface method caller trick? Is the naming for these two classes okay?

**Comment by modmuss50** [comment]
Im not a fan of the piston fix, what if some other block has a similar issue? (in the future possibly)

Is there a sane way to crash the game if someone accesses `BlockEntityRenderDispatcher.INSTANCE` too early?

**Comment by Draylar** [comment]
I'm not sure if there _are_ any alternatives to the piston fix. The offending method and field are unique to the piston BER; there's no single patch that would cover all possible issues. It's a rather unique implementation regardless, so I would guess we'd see changes to `BlockEntityRenderDispatcher` before we saw another BER that accesses `MinecraftClient#blockRenderManager`.

As long as `BlockEntityRenderDispatcher.INSTANCE` is open, there's a chance for a crash with pistons. The only other major way to get around this, ignoring a mixin to the piston BER, is by re-ordering how things are loaded. The potential fixes I can think of are:
   - don't load mods till `MinecraftClient#blockRenderManager` is initialized (impossible)
   - move `MC#blockRenderManager` initialization to before mods are loaded (probably horrible idea)
   - mess with `BlockEntityRenderDispatcher.INSTANCE` accessibility (impossible / horrid idea)
   - don't register piston renderer till after `MC#blockRenderManager` is loaded (seems overkill)
   - leave crash and tell people not to touch `BlockEntityRenderDispatcher.INSTANCE`

I'm certainly open to any other options, but this is all I can think of at the current moment. 

> Is there a sane way to crash the game if someone accesses BlockEntityRenderDispatcher.INSTANCE too early?

Yes-- we _could_ inject at the front of `BlockEntityRenderDispatcher#init`, check if `MC#blockRenderManager` is null, and crash if it is.

**Comment by asiekierka** [comment]
>Is there a sane way to crash the game if someone accesses BlockEntityRenderDispatcher.INSTANCE too early?

Couldn't we look into that thingabob I was thinking about, where you can block certain classes from being loaded by a mod before Minecraft itself loads them? I think it's worth at least brainstorming about for in-dev use...

**Comment by Player3324** [comment]
I'd only add some API to loader that can lock and unlock loading of arbitrary named classes (throw in loadClass if locked). Some official debug mod (included in example mod's gradle) can then add safeguards against this and many other problems.

I don't think it's trivial to determine what loaded a class.

**Comment by Draylar** [comment]
At the moment, I'd say the first priority is just preventing the crash, with future-proofing being something we can deal with later in the pre-release cycle. If we do end up waiting on any sort of extra API, the fix could easily drag into the end of December.

From a non-FabricAPI perspective, devs have the following options to avoid the crash:
  - don't register BERs
  - duplicate the current PR setup
  - mixin and register BERs after vanilla dispatcher constructor finishes [similar to above]
  - apply piston fix themselves

At the very least, we should add in a simple crash patch early on, even if we end up keeping the existing renderer registry (which would actually work fine, but is probably least ideal). These things considered, I think the best approach is:
  - merge current PR (after cleanup/changes) as a quick patch [mixin + new registry]
  - wait for 1.15 release to avoid missing any other similar changes
  - revisit patch to determine whether it's still needed
       - if the patch is still needed, look at changing field accessibility
       - if Mojang fixed the issue themselves, remove the mixin

I don't think it would make much sense to spend extra time future-proofing this fix when we still haven't finished pre-releases. The worst that could happen would be:
  - Mojang introduces a similar block in 1.16 and we add 1 more patch mixin
  - 1-2 mods reference `INSTANCE` and have to change it after Fabric API introduces a fix

File: fabric-rendering-v0/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinPistonBlockEntityRenderer.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+@Mixin(PistonBlockEntityRenderer.class)
+public class MixinPistonBlockEntityRenderer {
+	@Mutable
+	@Shadow
+	@Final
+	private BlockRenderManager manager;
+
+	@Inject(at = @At("HEAD"), method = "method_3575")
+	private void initializeNullManager(BlockPos pos, BlockState state, MatrixStack matrixStack, VertexConsumerProvider vertexConsumerProvider, World world, boolean boolean_1, int int_1, CallbackInfo ci) {
+		if (manager == null) {
```
Comment (modmuss50) [inline fabric-rendering-v0/src/main/java/net/fabricmc/fabric/mixin/client/rendering/MixinPistonBlockEntityRenderer.java]: For now id do away with this, I think coming up with a propper solution is better than this.

**Review by modmuss50 — CHANGES_REQUESTED** [review]
Looks mostly ok, I dont think we need the piston fix. And bump the module versions please.

**Comment by modmuss50** [comment]
I have reworked your PR slighly and commited it for pre2 https://github.com/FabricMC/fabric/commit/a1ff8ff878a834ff1f05aeb3689dd7e2daec8576

Thanks for this, hopefully it fixes the issues for now.

**Pull request closed without merging**